### DISCUSSION TITLE
LAN access with WSL2

### DISCUSSION THREAD

**Discussion by wickedwzrd** [Difegue/LANraragi#378] [Q&A] [answered]
**LRR Version and OS**
LRR Version: v.0.7.5 - Underground
OS: Windows 10 v2004

**Bug Details**
I am unable to access the reader on my LAN. I currently have LRR running on my home server and I tried to create an inbound rule for `Karen.exe` on any port and specifically 3000, but was still unable to access it from any device on my LAN.

I was scouring the closed issues and could only find that Mojolicious default config sets network to `http://*:3000`.

I would assume that LRR should work out of box over LAN so I am unsure as to what the issue may be.

**Matching Logs**
```
Executing the following command on WSL: export LRR_NETWORK=http://*:3000 && export LRR_DATA_DIRECTORY='/mnt/h/Comics/test' && cd /home/koyomi/lanraragi && rm -f script/hypnotoad.pid && mkdir -p log && mkdir -p content && mkdir -p database && sysctl vm.overcommit_memory=1 && redis-server /home/koyomi/lanraragi/tools/build/docker/redis.conf --dir '/mnt/h/Comics/test'/ --daemonize yes && perl ./script/launcher.pl -f ./script/lanraragi
vm.overcommit_memory = 1


ｷﾀ━━━━━━(ﾟ∀ﾟ)━━━━━━!!!!!
[LANraragi] [info] LANraragi 0.7.5 started. (Production Mode)
[LANraragi] [info] Plugin Detected: Chaika.moe
[LANraragi] [info] Plugin Detected: Tag Copier
[LANraragi] [info] Plugin Detected: Date Added
[LANraragi] [info] Plugin Detected: E-Hentai
[LANraragi] [info] Plugin Detected: eze
[LANraragi] [info] Plugin Detected: Hdoujin
[LANraragi] [info] Plugin Detected: koromo
[LANraragi] [info] Plugin Detected: Mayriad's EH Master Script
[LANraragi] [info] Plugin Detected: nHentai
[LANraragi] [info] Script Detected: Migrate Favorite Tags
[LANraragi] [info] Script Detected: Subfolders to Categories
[LANraragi] [info] Script Detected: Source Finder
[LANraragi] [info] Downloader Detected: Chaika.moe Downloader
[LANraragi] [info] Downloader Detected: E*Hentai Downloader
Killing process 30 from ./script/minion.pid
[LANraragi] [info] Successfully connected to Minion database.
[Minion] [info] Starting new Minion worker in subprocess with 8 parallel jobs.
[Minion] [info] Minion worker 30 started
Killing process 31 from ./script/shinobu.pid
[Minion] [info] Warming up search cache...
Logging to "log/production.log" is DEPRECATED at /home/koyomi/lanraragi/script/../lib/LANraragi/Utils/Generic.pm line 131.
[Minion] [info] Done!
[LANraragi] [info] Routing done! Ready to receive requests.
Logging to "log/production.log" is DEPRECATED at /usr/local/share/perl5/site_perl/Mojo/Server/Hypnotoad.pm line 69.
Web application available at http://127.0.0.1:3000
[Shinobu] [info] Shinobu File Watcher started.
[Shinobu] [info] Working dir is /home/koyomi/lanraragi
[Shinobu] [info] Scanning content folder for changes...
[Shinobu] [info] Adding watcher to content folder /mnt/h/Comics/test
[Shinobu] [info] All done! Now dutifully watching your files.```

**Comment by wickedwzrd** [comment]
Another note to add is that I am running LRR with WSL v2. I just tested switching back to WSL v1 and I have no issues connecting to the server. It asked me to allow perl access to private/public networks with the file location being `%appdata%\roaming\lanraragi\distro\rootfs\usr\bin\perl`

I noticed that when running WSL v2, LRR creates an ext4 image, while WSL v1 create a rootfs folder under the Distro folder.

**Comment by Difegue** [comment]
WSL2 currently has some issues related to localhost access by other machines: https://github.com/microsoft/WSL/issues/4636

For now I'd recommend you stick to WSL1 until said issues are fixed by Microsoft. 
Although you can try to hack around the problems if you really want to: https://stackoverflow.com/questions/61002681/connecting-to-wsl2-server-via-local-network/64986363

**Comment by huntrshado** [comment]
Seems like a firewall issue. Try creating an inbound rule for all programs to allow connection on port 3000 and see if it starts working.

I think it specifically has to do with making your firewall rule only allow Karen.exe, because when I test on my own device and configure the rule to allow only Karen.exe through port 3000, it breaks and I also become unable to access my server.

e: as you both confirmed, it is a WSL2 bug. But if you configure an open rule on port 3000 while on WSL2, it works fine.

**Comment by wickedwzrd** [comment]
@huntrshado I noticed that while using WSL1, it creates a firewall rule for perl and not the `karen.exe`. However since WSL1 is easier to access the root folder so that I can edit my theme I will stick to that for now.

**Comment by Difegue** [comment]
Since there's no real fixable issue on my side, I'll be converting this to a discussion so that others can see it if they run into trouble w. WSL2 installs. Glad you got it sorted out!

**Comment by Zero-G-Sys** [comment, marked as answer]
Opening the port on the firewall wasn't enough for me. I found the solution in this link:
https://stackoverflow.com/questions/61002681/connecting-to-wsl2-server-via-local-network

Running this command (as administrator) (and having the port open on the firewall already) made me able to access LANraragi from my local network when using WSL2:
`netsh interface portproxy add v4tov4 listenaddress=0.0.0.0 listenport=3000 connectaddress=localhost connectport=3000`

But, then I restarted and wanted to see if the port proxy forward was still active. Running the command
`netsh interface portproxy show all`
showed that it was indeed still active, so I ran LANraragi again and... didn't have access from my local network.

Turns out that you can't have the port proxy forward active when LANraragi starts, you can only activate it once it has started. I created a couple of bat files to run before and after starting LANraragi, this probably can be integrated into Karen.

Bat before running LANraragi (Should return empty):
```
@echo off
netsh interface portproxy delete v4tov4 listenaddress=0.0.0.0 listenport=3000
netsh interface portproxy show all
pause
```

Bat after running LANraragi (and it's serving locally):
```
@echo off
netsh interface portproxy add v4tov4 listenaddress=0.0.0.0 listenport=3000 connectaddress=localhost connectport=3000
netsh interface portproxy show all
pause
```

Should return:
```
Listen on ipv4:             Connect to ipv4:

Address         Port        Address         Port
--------------- ----------  --------------- ----------
0.0.0.0         3000        localhost       3000
```

**Reply by a942533331** [reply]
thank you bro,  its worked

**Reply by eao** [reply]
Thank you for the solution.

To get this to happen automatically, here is a .ps1 file I wrote to run at startup:

```
Start-Sleep -Seconds 5
netsh interface portproxy delete v4tov4 listenaddress=0.0.0.0 listenport=3000
Start-Sleep -Seconds 10
C:\Users\User\AppData\Roaming\LANraragi\Karen.exe
Start-Sleep -Seconds 20
netsh interface portproxy add v4tov4 listenaddress=0.0.0.0 listenport=3000 connectaddress=localhost connectport=3000
```

Obviously replace the path to Karen.exe with your own path.  The sleep commands are probably overly conservative but I had failures with lower values and I don't really care if it takes 30 seconds for it to start up. 

Then, to get this script to run invisibly at startup, go to Task Scheduler, make a new task, and check "Run with highest privileges".  Make the trigger logon, and the action "Start a program".  For the program, just type `cmd`, and for the "Add arguments" box, put in `cmd /c start /min "" powershell -WindowStyle Hidden -ExecutionPolicy Bypass -File "C:\Users\User\Documents\batchest\lrr.ps1"`.  Obviously replacing the path to the .ps1 file with your own path.

**Reply by owenk10** [reply]
> 
> Then, to get this script to run invisibly at startup, go to Task Scheduler, make a new task, and check "Run with highest privileges". Make the trigger logon, and the action "Start a program". For the program, just type `cmd`, and for the "Add arguments" box, put in `cmd /c start /min "" powershell -WindowStyle Hidden -ExecutionPolicy Bypass -File "C:\Users\User\Documents\batchest\lrr.ps1"`. Obviously replacing the path to the .ps1 file with your own path.

Thank you for posting the `.ps1` script, it worked! However, I can't get the Task Scheduler to work.. 

Might be caused by me not knowing how to create a .ps1 script, i followed the Microsoft Learn forum to create one based on your script. 

In that tutorial, they said to set the `ExecutionPolicy` to `RemoteSigned`, would that affect the `-ExecutionPolicy Bypass`? If so, how could it change it?

**Reply by eao** [reply]
To make the .ps1 script, basically, all you should need to do is open a text editor such as Notepad, copy and paste the 6 lines from my post into it, then save it as "lrr.ps1".  Not that the filename necessarily needs to be "lrr," but whatever.  If that is posing problems, I have uploaded a zip file that has lrr.ps1 in it, so you can just use this:  [take lrr out of this zip file and put it somewhere.zip](https://github.com/user-attachments/files/16165025/take.lrr.out.of.this.zip.file.and.put.it.somewhere.zip)

As for getting Task Scheduler to work, this is a bit of the blind leading the blind, as I just copied and pasted stuff from stackoverflow until it worked.  (https://stackoverflow.com/questions/1802127/how-to-run-a-powershell-script-without-displaying-a-window)  Assuming your lrr.ps1 file is fine, I don't know what would be causing your issue.  If you post a picture of your Task Scheduler window I might be able to help you figure out what's going wrong.

Or, if you don't care about it running invisibly, you could just try having the program be `powershell` and the arguments be `-File "C:\Your\Path\Goes\Here\This\Is\Just\An\Example\Path\lrr.ps1"`.

**Comment by ilevalser** [comment]
So this is also the reason why clash tun mode can no longer proxy LRR? The two methods I got in [#739](https://github.com/Difegue/LANraragi/discussions/739) are now not applicable. Is there any way to make LRR go through the proxy?
